### PR TITLE
vmalert: fix `externalURL` function in `-rule.templates` after config…

### PR DESCRIPTION
**Pull request by Haleygo**
`externalURL` function was available for `-rule.templates` templates in first init, but lost after config reload.

### CODE REVIEW

File: app/vmalert/rule/alerting.go [hunk 1]
```diff
@@ -313,11 +313,8 @@ func (ar *AlertingRule) execRange(ctx context.Context, start, end time.Time) ([]
 	}
 	var result []prompbmarshal.TimeSeries
 	holdAlertState := make(map[uint64]*notifier.Alert)
-	qFn := func(_ string) ([]datasource.Metric, error) {
-		return nil, fmt.Errorf("`query` template isn't supported in replay mode")
-	}
 	for _, s := range res.Data {
-		ls, as, err := ar.expandTemplates(s, qFn, time.Time{})
+		ls, as, err := ar.expandTemplates(s, nil, time.Time{})
```
Comment (f41gh7) [inline app/vmalert/rule/alerting.go:317]: This commit contains hidden change - it removes `query` function from `query_range` aka `vmalert-replay` mode. 
 
 WDYT of replacing it with `no-op` function? Because after this change, it may require to rewrite all alerting rules, before passing to the `replay-mode`. Which makes it hard to use.

Comment (Haleygo) [reply app/vmalert/rule/alerting.go:317]: The behavior doesn't change, query func by default return a similar error and is no-op.
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3513eb04bdc78ee9983156dd970f030884424b8b/app/vmalert/templates/template.go#L149-L151
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3513eb04bdc78ee9983156dd970f030884424b8b/app/vmalert/notifier/alert.go#L154-L158

Comment (f41gh7) [reply app/vmalert/rule/alerting.go:317]: I'll create an issue for it. Looks like it's not possible to use replay mode with `query` function.

File: app/vmalert/templates/template.go [hunk 2]
```diff
@@ -198,15 +159,16 @@ func FuncsWithQuery(query QueryFn) textTpl.FuncMap {
 	}
 }
 
-// FuncsWithExternalURL returns a function map that depends on externalURL value
-func FuncsWithExternalURL(externalURL *url.URL) textTpl.FuncMap {
+// funcsWithExternalURL returns a function map that depends on externalURL value
+func funcsWithExternalURL(externalURL string) textTpl.FuncMap {
+	eu, _ := url.Parse(externalURL)
```
Comment (f41gh7) [inline app/vmalert/templates/template.go:164]: I'd suggest to assert with logger.Panicf() if parsing error is not nil in order to mitigate possible bugs with `nil pointer deref`.

Comment (Haleygo) [reply app/vmalert/templates/template.go]: Made the function more reliable [here](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7773/commits/1f4945b86f49f5677e17a5b498fd04f008508761).

File: app/vmalert/main.go [hunk 3]
```diff
@@ -95,9 +95,14 @@ func main() {
...
 	logger.Init()
 
-	err := templates.Load(*ruleTemplatesPath, true)
+	eu, err := getExternalURL(*externalURL)
 	if err != nil {
-		logger.Fatalf("failed to parse %q: %s", *ruleTemplatesPath, err)
+		logger.Fatalf("failed to init external.url %q: %s", *externalURL, err)
+	}
+
+	err = templates.Load(*ruleTemplatesPath, *externalURL)
```
Comment (hagen1778) [inline app/vmalert/main.go:103]: Passing raw `*externalURL` here would produce different results to `getExternalURL `.
Why not passing `eu *url.URL` instead?

File: app/vmalert/main.go [hunk 4]
```diff
@@ -299,12 +299,9 @@ func getAlertURLGenerator(externalURL *url.URL, externalAlertSource string, vali
 		"tpl": externalAlertSource,
 	}
 	return func(alert notifier.Alert) string {
-		qFn := func(_ string) ([]datasource.Metric, error) {
```
Comment (hagen1778) [inline app/vmalert/main.go:302]: why this change was made? the error message here looks more precise and clear to user than internal message

File: app/vmalert/rule/alerting.go [hunk 5]
```diff
@@ -313,11 +313,8 @@ func (ar *AlertingRule) execRange(ctx context.Context, start, end time.Time) ([]
 	}
 	var result []prompbmarshal.TimeSeries
 	holdAlertState := make(map[uint64]*notifier.Alert)
-	qFn := func(_ string) ([]datasource.Metric, error) {
-		return nil, fmt.Errorf("`query` template isn't supported in replay mode")
```
Comment (hagen1778) [inline app/vmalert/rule/alerting.go:317]: same about message context is here

File: app/vmalert/main.go (hunk 3 above)
Comment (Haleygo) [reply app/vmalert/main.go]: Good point, fixed in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7773/commits/4f32b5a9cbd837881ca37bd7dc4f8a4271cae346.

File: app/vmalert/main.go (hunk 4 above)
Comment (Haleygo) [reply app/vmalert/main.go:302]: I don't see the need to define an invalid qFn() everywhere, as the error will be wrapped in the following line:
```
return nil, fmt.Errorf("`query` template isn't supported for alert source template")
``` 
The error messages from the past and present are basically the same.
```
before: can not exec source template errors(1): key "tpl", template "xxx": error evaluating annotation template: template: :1:330: executing "" at <query "sort_desc(process_resident_memory_bytes)">: error calling query: `query` template isn't supported for alert source template

now: cannot template alert source: errors(1): key "tpl", template "xxxx": error evaluating annotation template: template: :1:330: executing "" at <query "sort_desc(process_resident_memory_bytes)">: error calling query: cannot execute query "sort_desc(process_resident_memory_bytes)": query is not available in this context
```

Comment (f41gh7) [reply app/vmalert/main.go:302]: error `"`query` template isn't supported in replay mode"` has more context and it's easier to understand for users.

Comment (Haleygo) [reply app/vmalert/main.go]: ok, reverted, please check

**Comment by f41gh7** [comment]
Thanks for contribution!
